### PR TITLE
[ Frontend ] Multiprocessing for OpenAI Server with `zeromq`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
RFC: https://github.com/vllm-project/vllm/issues/6797

SUMMARY:
* Use multiprocessing for OpenAI Server such that it does not contend with `AsyncLLMEngine`
* Use `zeromq` for networking between OAI process and `AsyncLLMEngine`
* Currently uses pickling for python serialization, to update this to use protobufs in separate PR
* Gives ~20% TPOT improvement for QPS 10 for `llama-8b-instruct` on H100 (550 input tokens, 150 output tokens)

TODO:

- [x] enabling aborting
- [x] enable do_log_stats
- [x] error handling
- [x] refactor to use single socket
- [x] use open port
- [x] turn on the other OAI endpoints
- [x] add flag to disable mp
- [x] tokenizer 
- [x] guided decoding
- [x] tracing
- [x] make CI green

PERFORMANCE:
* `zeromq` based protocol is delivering speedup in high QPS scenario:
*  Llama-3-8B-Instruct, 1xH100, 550 input tokens | 150 output tokens, 10 QPS, chunked prefill

| Metric | `main` | `grpc` | `zmq + pickle` (this PR) | `zmq + protobuf` |
| - | - | - | - | - |
| TPOT (mean) | 70.74 | 64.15 | 57.35 | 54.54 |
| TTFT (mean) | 687.69 | 450.05 | 298.09 | 269.01 |

Clear win + clear we should move forward with `zeromq`

---

Benchmark script for reference:

```bash
MODEL="meta-llama/Meta-Llama-3-8B-Instruct"
TOTAL_SECONDS=60
QPS_RATES=("10")

for QPS in ${QPS_RATES[@]}; do
    NUM_PROMPTS=$((TOTAL_SECONDS * QPS))
    echo "===== RUNNING NUM_PROMPTS = $NUM_PROMPTS QPS = $QPS ====="

    python3 benchmarks/benchmark_serving.py \
        --model $MODEL \
        --dataset-name sonnet --sonnet-input-len 550 --sonnet-output-len 150 --dataset-path benchmarks/sonnet.txt \
        --num-prompts $NUM_PROMPTS --request-rate $QPS
done
```

--- 

FOLLOW UP WORK:
* Refactor how logits processors are handled. They should not belong to the API layer anymore, rather below the engine. Currently there are hacks around making it serializable. 
* Switch to using a `protobuf` communication layer for the RPC Server
* Move detokenization to the OAI layer
* Profile using `N` uvicorn processes + expose a flag
* Support embedding models via RPC

### CODE REVIEW

File: vllm/entrypoints/openai/rpc/__init__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from enum import Enum
+
+from vllm.inputs import PromptInputs
+from vllm.lora.request import LoRARequest
+from vllm.prompt_adapter.request import PromptAdapterRequest
+from vllm.sampling_params import SamplingParams
+
+VLLM_GENERATE_RPC_PATH = "tcp://localhost:5570"
+VLLM_GET_DATA_RPC_PATH = "tcp://localhost:5571"
+VLLM_IS_READY_RPC_PATH = "tcp://localhost:5572"
```
Comment (mgoin) [inline vllm/entrypoints/openai/rpc/__init__.py:12]: At the very least we should be choosing randomly available ports above the "user space" of the 1000s port range. See https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.bind_to_random_port

File: vllm/entrypoints/openai/rpc/server.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        async for request_output in results_generator:
+            self.generate_socket.send_multipart([
+                identity,
+                pickle.dumps(request_output, pickle.HIGHEST_PROTOCOL)
+            ])
+
+    async def run_loop(self):
+        # Notify the RPC client that we are ready to recieve requests.
+        await self.is_ready_socket.send_string("Ready!")
```
Comment (mgoin) [inline vllm/entrypoints/openai/rpc/server.py:70]: This ready string should be a global shared constant

File: vllm/entrypoints/openai/rpc/server.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+            # We need to keep around a strong reference to the task,
+            # to avoid the task disappearing mid-execution as running tasks
+            # can be GC'ed. Below is a common "fire-and-forget" tasks
+            # https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task
+            if task is not None:
+                running_tasks.add(task)
+                task.add_done_callback(running_tasks.discard)
+
+        # TODO: Do I need to close the generate / get_data sockets?
```
Comment (mgoin) [inline vllm/entrypoints/openai/rpc/server.py:99]: I think it's best to close sockets explicitly if possible per https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.close

> If this is not called, the socket will automatically be closed when it is garbage collected, in which case you may see a ResourceWarning about the unclosed socket.

**Comment by WoosukKwon** [comment]
Awesome! This is really great!

File: vllm/entrypoints/openai/api_server.py [hunk 4]
```diff
@@ -62,12 +67,12 @@ async def lifespan(app: fastapi.FastAPI):
     async def _force_log():
         while True:
             await asyncio.sleep(10)
-            await engine.do_log_stats()
+            await backend.do_log_stats()
```
Comment (njhill) [inline vllm/entrypoints/openai/api_server.py:70]: I think this should just move into AsyncLLMEngine right? (perhaps whether it's started controlled by some arg)

File: vllm/entrypoints/openai/serving_completion.py [hunk 5]
```diff
@@ -93,10 +93,10 @@ async def create_completion(self, request: CompletionRequest,
                 prompt_adapter_request,
             ) = self._maybe_get_adapters(request)
 
-            tokenizer = await self.engine.get_tokenizer(lora_request)
+            tokenizer = await self.vllm_backend.get_tokenizer(lora_request)
```
Comment (njhill) [inline vllm/entrypoints/openai/serving_completion.py:96]: Presumably the idea is that these would never actually be remote calls and instead live within the "client" layer of the remote VLLMBackend impl (get_tokenizer, get_decoding_config, is_tracing_enabled, etc).

File: vllm/utils.py [hunk 6]
```diff
@@ -315,7 +322,13 @@ async def producer(i: int, iterator: AsyncIterator[T]):
     async def consumer():
         try:
             while not all(finished) or not queue.empty():
```
Comment (njhill) [inline vllm/utils.py:324]: I think we can eliminate the `finished` list and just have a count that's decremented in the consumer when it receives each `ProducerFinished`?

File: vllm/entrypoints/openai/rpc/client.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        # Connect to RPC socket for Request-Reply pattern,
+        # Note that we use DEALER to enable asynchronous communication
+        # to enable streaming.
+        socket = self.context.socket(zmq.constants.DEALER)
+        socket.connect(VLLM_GENERATE_RPC_PATH)
+
+        # Send GenerateRequest to the RPC Server.
+        await socket.send_multipart([
+            pickle.dumps(
```
Comment (njhill) [inline vllm/entrypoints/openai/rpc/client.py:82]: As @robertgshaw2-neuralmagic suggested, let's use msgspec?

**Review by njhill — COMMENTED** [review]
Thanks @joerunde @robertgshaw2-neuralmagic this looks great so far!

File: vllm/entrypoints/openai/serving_completion.py (hunk 5 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/serving_completion.py:96]: We we should just pull over all the state we need at the start once

File: vllm/entrypoints/openai/rpc/client.py (hunk 7 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/rpc/client.py:82]: Let's separate the messaging protocol optimizations to a separate PR

File: vllm/entrypoints/openai/api_server.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/api_server.py:70]: Fine with me if it works

File: vllm/entrypoints/openai/serving_completion.py (hunk 5 above)
Comment (njhill) [reply vllm/entrypoints/openai/serving_completion.py:96]: Yes exactly, no problem with that being common logic with the single-process case too.

Though get_tokenizer might need to be slighlty different, since we may need to load lora-speciic tokenizers on the fly.

File: vllm/entrypoints/openai/rpc/__init__.py (hunk 1 above)
Comment (joerunde) [reply vllm/entrypoints/openai/rpc/__init__.py:12]: @mgoin yeah that'd be better. Any idea how we'd notify the clients what port to connect to in that case?

The server that would be calling `.bind_to_random_port()` is in a different process than the openai server that needs to connect clients to it

Comment (youkaichao) [reply vllm/entrypoints/openai/rpc/__init__.py:12]: the the openai server can listen first, pass the ports to the server process, and then the server just connects to it?

File: vllm/engine/protocol.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+from vllm.lora.request import LoRARequest
+from vllm.outputs import EmbeddingRequestOutput, RequestOutput
+from vllm.pooling_params import PoolingParams
+from vllm.prompt_adapter.request import PromptAdapterRequest
+from vllm.sampling_params import SamplingParams
+from vllm.sequence import SamplerOutput
+
+
+@runtime_checkable
+class VLLMBackend(Protocol):
```
Comment (simon-mo) [inline vllm/engine/protocol.py:18]: Naming wise, I think this the protocol for `AsyncEngineClient` or `AsyncEngine`. Also I prefer no vLLM in the class name. The entire package is named under vLLM. Backend is ambiguous because we already use it for hardware backends, and attention backends.

File: vllm/entrypoints/openai/api_server.py [hunk 9]
```diff
@@ -72,6 +77,60 @@ async def _force_log():
...
+    # Context manager to handle backend lifecycle
+    # Ensures everything is shutdown and cleaned up on error/exit
+    global engine_args
+    engine_args = AsyncEngineArgs.from_cli_args(args)
+
+    # Backend itself still global for the silly lil' health handler
+    global backend
+
+    # First need to determine if this is an embeddings model
+    # (no remote backend for those)
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:91]: why?

File: vllm/entrypoints/openai/api_server.py [hunk 10]
```diff
@@ -72,6 +77,60 @@ async def _force_log():
...
+    model_config = ModelConfig(model=args.model,
+                               tokenizer=args.tokenizer,
+                               tokenizer_mode="auto",
+                               trust_remote_code=False,
+                               seed=0,
+                               dtype="float16")
+    if model_config.embedding_mode or args.disable_frontend_multiprocessing:
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:98]: make this a util? 
`def model_is_embedding(model, tokenizer)`

**Review by simon-mo — COMMENTED** [review]
reviewed non `rpc` code. recommend the rename from vllm_backend back to engine or engine client.

File: vllm/entrypoints/openai/api_server.py (hunk 9 above)
Comment (joerunde) [reply vllm/entrypoints/openai/api_server.py:91]: cc @robertgshaw2-neuralmagic 

I think because it'd be more work to hook those up, and we're not worried about the throughput improvements for embeddings atm

Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/api_server.py:91]: Planning to do follow to enable RPC for Embedding

File: vllm/entrypoints/openai/rpc/server.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
+import asyncio
+import signal
+from typing import Any, Coroutine
+
+import cloudpickle
```
Comment (simon-mo) [inline vllm/entrypoints/openai/rpc/server.py:5]: need this in requirements txt

File: vllm/entrypoints/openai/rpc/server.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+from typing_extensions import Never
+
+from vllm import AsyncEngineArgs, AsyncLLMEngine
+from vllm.entrypoints.openai.rpc import (VLLM_RPC_HEALTHY_STR,
+                                         VLLM_RPC_SUCCESS_STR, RPCAbortRequest,
+                                         RPCGenerateRequest, RPCUtilityRequest)
+from vllm.logger import init_logger
+from vllm.usage.usage_lib import UsageContext
+
+logger = init_logger('vllm.entrypoints.openai.rpc.server')
```
Comment (simon-mo) [inline vllm/entrypoints/openai/rpc/server.py:17]:
```suggestion
logger = init_logger(__name__)
```

File: vllm/engine/protocol.py (hunk 8 above)
Comment (joerunde) [reply vllm/engine/protocol.py:18]: Yeah `AsyncEngine` is pretty spot on

Comment (robertgshaw2-neuralmagic) [reply vllm/engine/protocol.py:18]: I think we should call it a client

**Review by simon-mo — COMMENTED** [review]
I think any non generation request can be serialized by `pickle` instead of `cloudpickle`, which is preferred performance wise, and prevent accidentally leaking large data structure.

**Review by simon-mo — COMMENTED** [review]
LGTM over all. This is a good direction. Can we check one edge case? 
* Send request generate 2K toks
* Drop the connection 
* Check the `generate` coroutine is properly cancelled, and the abort it called, and there is no leakage of coroutines in both client and server (by printing out something...)

File: vllm/entrypoints/openai/api_server.py (hunk 9 above)
Comment (simon-mo) [reply vllm/entrypoints/openai/api_server.py:91]: leave a comment saying future work plz.

File: vllm/engine/protocol.py (hunk 8 above)
Comment (joerunde) [reply vllm/engine/protocol.py:18]: I don't think we should call it a client, this is the name of the interface that both the `AsyncLLMEngine` itself and the client to the backend process running the `AsyncLLMEngine` implement.

I'd be up for renaming like `RPCClient` -> `AsyncLLMEngineClient` and `RPCServer` -> `AsyncLLMEngineServer` if we wanna be consistent with these names

File: vllm/engine/protocol.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    async def get_decoding_config(self) -> DecodingConfig:
+        """Get the decoding configuration of the vLLM engine."""
+
+    async def get_tokenizer(
+        self,
+        lora_request: Optional[LoRARequest] = None,
+    ) -> PreTrainedTokenizer:
+        """Get the appropriate Tokenizer for the request"""
+
+    async def is_tracing_enabled(self) -> bool:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/protocol.py:73]: @joerunde why are these `pass`?

**Comment by robertgshaw2-neuralmagic** [comment]
> LGTM over all. This is a good direction. Can we check one edge case?
> 
> * Send request generate 2K toks
> * Drop the connection
> * Check the `generate` coroutine is properly cancelled, and the abort it called, and there is no leakage of coroutines in both client and server (by printing out something...)

Do you want this in automation?

File: vllm/entrypoints/openai/rpc/server.py (hunk 11 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/rpc/server.py:5]: I can explictily add it, but it seems to be installed already

**Comment by robertgshaw2-neuralmagic** [comment]
| Branch | 2 |  6 | 10 | 14 |
| - | - | - | - | - |

**Review by simon-mo — APPROVED** [review]
:shipit:

File: vllm/entrypoints/openai/rpc/client.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+                pass
+            else:
+                raise ValueError(error_message)
+
+        return data
+
+    async def _send_one_way_rpc_request(self, request: RPC_REQUEST_TYPE,
+                                        error_message: str):
+        """Send one-way RPC request to trigger an action."""
+        with self.socket() as socket:
```
Comment (youkaichao) [inline vllm/entrypoints/openai/rpc/client.py:91]: I'm trying to understand this. We create a socket every time we call this function?

File: vllm/entrypoints/openai/api_server.py [hunk 15]
```diff
@@ -72,6 +86,52 @@ async def _force_log():
...
+        return
+
+    # Otherwise, use the multiprocessing AsyncLLMEngine.
+    else:
+        # Start RPCServer in separate process (holds the AsyncLLMEngine).
+        port = get_open_port(envs.VLLM_RPC_PORT)
+        rpc_server_process = Process(target=run_rpc_server,
+                                     args=(engine_args,
+                                           UsageContext.OPENAI_API_SERVER,
+                                           port))
```
Comment (youkaichao) [inline vllm/entrypoints/openai/api_server.py:115]: why do we need a new env var here? isn't `get_open_port()` enough?

**Review by youkaichao — COMMENTED** [review]
Thanks for the great work! I just left some nit comments that I don't understand. If you have time, could you please briefly draw some diagrams to show how many processes do we have use and how do they interact with each other?

File: tests/entrypoints/openai/test_disable_mp.py [hunk 16]
```diff
@@ -0,0 +1,715 @@
+"""
+Repeat of tests in test_completion.py with the non-mp backend.
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply tests/entrypoints/openai/test_disable_mp.py:2]: Because we need to modify default server args which is a fixture for the whole file so unfortunately I think this has to be a separate file

File: vllm/entrypoints/openai/api_server.py (hunk 15 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/api_server.py:115]: I would be open to reverting this. Will just need to check how it interacts with tp

File: vllm/entrypoints/openai/rpc/client.py (hunk 14 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/rpc/client.py:91]: Yes. I believe this is a common paradigm for zeromq based clients

also, this function is mostly used in setup

Comment (robertgshaw2-neuralmagic) [reply vllm/entrypoints/openai/rpc/client.py:91]: Also, this does not mean a new Unix socket is created. This is a zeromq socket. how zeromq handles this internally is opaque to us
